File: week3_hash_tables/2_hash_chains/hash_chains.py
# python3

class Query:

    def __init__(self, query):
        self.type = query[0]
        if self.type == 'check':
            self.ind = int(query[1])
        else:
            self.s = query[1]

class Node:
    def __init__(self,string,prev=None,next=None):
        self.string = string
        self.prev = prev
        self.next = next

class Chain:
    
    def __init__(self):
        self.head = None
    
    
    def find(self,string):
        it = self.head
        while(it != None):
            if (it.string == string): break
            it = it.next
        return it
        
    
    def add(self,string):
        it = self.find(string)
        if (it!=None): return
        
        newnode = Node(string,None,self.head)
        if self.head:
            self.head.prev = newnode
        self.head = newnode
        
    
    def delete(self,string):
        it = self.find(string)
        if(it != None):
            if (it == self.head):
                self.head = it.next
            else:
                it.prev.next = it.next
            if (it.next !=None):
                it.next.prev = it.prev
    


class QueryProcessor:
    _multiplier = 263
    _prime = 1000000007

    def __init__(self, bucket_count, file=None):
        self.bucket_count = bucket_count
        # store all strings in one list
        self.elems = list(Chain() for _ in range(self.bucket_count))
        self.infile = file

    def _hash_func(self, s):
        ans = 0
        for c in reversed(s):
            ans = (ans * self._multiplier + ord(c)) % self._prime
        return ans % self.bucket_count

    def write_search_result(self, was_found):
        print('yes' if was_found else 'no')

    def write_chain(self, chain):
        print(' '.join(chain))

    def read_query(self):
        return Query(input().split())
        #return Query(self.infile.readline().split())

    def process_query(self, query):
        if query.type == "check":
            words = []
            it = self.elems[query.ind].head
            while (it!=None):
                words.append(it.string)
                it = it.next
            self.write_chain(words)
            
        else:
            ind = self._hash_func(query.s)
            if query.type == 'find':
                self.write_search_result(self.elems[ind].find(query.s) != None)
            elif query.type == 'add':
                    self.elems[ind].add(query.s)
            else:
                self.elems[ind].delete(query.s)


    def process_queries(self):
        n = int(input())
        #n = int(self.infile.readline())
        for i in range(n):
            self.process_query(self.read_query())

if __name__ == '__main__':
    #cmds = open('/Users/rajgopalan/data_sc_sample_projects/data_structures/week3_hash_tables/2_hash_chains/tests/06','r')
    #line = cmds.readline()
    #bucket_count = int(line)
    bucket_count = int(input())
    proc = QueryProcessor(bucket_count)
    proc.process_queries()
    #cmds.close()

File: week3_hash_tables/2_hash_chains/hash_chains.cpp
#include <iostream>
#include <string>
#include <vector>
#include <algorithm>
#include<unordered_set>

using std::string;
using std::vector;
using std::cin;
using std::unordered_set;

struct Query {
    string type, s;
    size_t ind;
};

class QueryProcessor {
    int bucket_count;

    struct hash_func{
        int bucket_count;
        hash_func(): bucket_count(1000){}
        hash_func(int count): bucket_count(count){}

        const size_t operator()(const string & s) const {
            static const size_t multiplier = 263;
            static const size_t prime = 1000000007;
            unsigned long long hash = 0;
            for (int i = static_cast<int> (s.size()) - 1; i >= 0; --i)
                hash = (hash * multiplier + s[i]) % prime;
            return (const size_t)(hash % bucket_count);
        }
    };

    unordered_set<string, hash_func> elems;

public:
    explicit QueryProcessor(int bucket_count): bucket_count(bucket_count) {       
        elems = unordered_set<string, hash_func> (bucket_count,hash_func(bucket_count));
    }
    

    Query readQuery() const {
        Query query;
        cin >> query.type;
        if (query.type != "check")
            cin >> query.s;
        else
            cin >> query.ind;
        return query;
    }

    void writeSearchResult(bool was_found) const {
        std::cout << (was_found ? "yes\n" : "no\n");
    }

    void processQuery(const Query& query) {
        
        if (query.type == "check") {
            for (auto it = elems.begin(query.ind); it!= elems.end(query.ind); it++)
                std::cout<< *it <<" ";
            std::cout << "\n";
        } 
        
        else {
            auto it = elems.find(query.s);
            
            if (query.type == "find"){
                writeSearchResult(it != elems.end());
            }
            else if (query.type == "add") {
                if (it == elems.end())
                    elems.insert(query.s);
            } else if (query.type == "del") {
                elems.erase(query.s);
            }
        }
    }

    void processQueries() {
        int n;
        cin >> n;
        for (int i = 0; i < n; ++i)
            processQuery(readQuery());
    }
};

int main() {
    std::ios_base::sync_with_stdio(false);
    int bucket_count;
    cin >> bucket_count;
    QueryProcessor proc(bucket_count);
    proc.processQueries();
    return 0;
}
